src/Microsoft.AspNetCore.Mvc.Core/Routing/ConsumesMatcherPolicy.cs

```diff
@@ -0,0 +1,242 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Formatters;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.Routing.Matchers;
+using Microsoft.AspNetCore.Routing.Patterns;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Mvc.Routing
+{
+    internal class ConsumesMatcherPolicy : MatcherPolicy, IEndpointComparerPolicy, INodeBuilderPolicy
+    {
+        internal const string Http415EndpointDisplayName = "415 HTTP Unsupported Media Type";
+        internal const string AnyContentType = "*/*";
+
+        // Run after HTTP methods, but before 'default'.
+        public override int Order { get; } = -100;
+
+        public IComparer<Endpoint> Comparer { get; } = new ConsumesMetadataEndpointComparer();
+
+        public bool AppliesToNode(IReadOnlyList<Endpoint> endpoints)
+        {
+            if (endpoints == null)
+            {
+                throw new ArgumentNullException(nameof(endpoints));
+            }
+
+            return endpoints.Any(e => e.Metadata.GetMetadata<IConsumesMetadata>()?.ContentTypes.Count > 0);
+        }
+
+        public IReadOnlyList<PolicyNodeEdge> GetEdges(IReadOnlyList<Endpoint> endpoints)
+        {
+            if (endpoints == null)
+            {
+                throw new ArgumentNullException(nameof(endpoints));
+            }
+
+            // The algorithm here is designed to be preserve the order of the endpoints
```

Comment:
This follows the structure of the HTTP method logic for building the nodes. I literally just copy-pasted the code and edited it since the comments are pretty applicable